src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -107,7 +106,9 @@ public bool ReceivedEmptyRequestBody
             }
         }
 
-        public bool CanReuse { get; private set; }
+        // We only want to reuse a stream that was not aborted and has completely finished writing.
+        // This ensures Http2OutputProducer.ProcessDataWrites is in the correct state to be reused.
+        public bool CanReuse => !_connectionAborted && HasResponseCompleted;
 
         protected override void OnReset()
```

Comment:
Would resetting the value to `false`  in OnReset accomplish the same goal?

Reply:
The line on 44 was effectively doing that before. The issue was that `!_connectionAborted && HasResponseCompleted` was being evaluated too early on a background thread and a `RST_STREAM` frame could still be processed and abort the stream before pooling. Now `CanReuse` is evaluated once in the main frame-processing loop right before it's pooled or not.